### PR TITLE
bugfix: Fix that cannot update bridge when restart pouchd

### LINKED ISSUE #2128: the host's network interface settings can not keep update after reset the default bridge network configurations
### Ⅰ. Issue Description
the host's network interface settings can not keep update after reset the default bridge network configurations by /etc/pouch/config.json 

1. the default bridge network settings info:
``` shell
pouch network inspect bridge

[
    {
        "Driver": "bridge",
        "IPAM": {
            "Config": [
                {
                    "Gateway": "192.168.5.1",
                    "IPRange": "192.168.5.1/24",
                    "Subnet": "192.168.5.1/24"
                }
            ],
            "Driver": "default"
        },
        "Id": "5702b5163abb6d6bbc7fd1a21189ce8da9c63f0e4bcaafd0a2b667ec0eec5b9a",
        "Name": "bridge",
        "Options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "p0",
            "com.docker.network.driver.mtu": "1500"
        }
    }
]
```

``` shell
ifconfig

p0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.5.1  netmask 255.255.255.0  broadcast 192.168.5.255
        ether 02:42:c0:a8:05:01  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

2. after reset the default bridge network settings

``` shell
cat /etc/pouch/config.json
{
  "network-config": {
    "bridge-config": {
      "bridge-name": "p0",
      "bip": "172.19.0.1/16",
      "mtu": 1500,
      "default-gateway": "172.19.0.1"
    }
  }
}
```

``` shell
systemctl restart pouch

pouch network inspect bridge
[
    {
        "Driver": "bridge",
        "IPAM": {
            "Config": [
                {
                    "Gateway": "172.19.0.1",
                    "IPRange": "172.19.0.1/16",
                    "Subnet": "172.19.0.1/16"
                }
            ],
            "Driver": "default"
        },
        "Id": "bd74ac1f038bd0b0b13652d95dcb436bd06819e1e2614194382fc1e1bfc89c08",
        "Name": "bridge",
        "Options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "p0",
            "com.docker.network.driver.mtu": "1500"
        }
    }
]

ifconfig 
p0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.5.1 (** Error! **) netmask 255.255.255.0  broadcast 192.168.5.255
        ether 02:42:c0:a8:05:01  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1.
2.
3.

### Ⅴ. Anything else we need to know?

i modify the configuration of network interface on host manually by ```ifconfig p0 172.19.0.1```, then the pouch network is working prety good. but, when i restart the server, the host's network interface gateway was reset to default bridge address to 192.168.5.1.

### Ⅵ. Environment:

- pouch version (use `pouch version`):
BuildTime:       2018-08-20T20:10:16+08:00
GitCommit:       1.0.0.rc1-244-g94baba40-dirty
GoVersion:       go1.10.3
KernelVersion:   3.10.0-862.9.1.el7.x86_64
Os:              linux
Version:         1.0.0-rc1
APIVersion:      1.24
Arch:            amd64

- OS (e.g. from /etc/os-release):
CentOS 7.5.1804
- Kernel (e.g. `uname -a`):
 3.10.0-862.9.1

- Install tools:
source make install

- Others:

### PR DESCRIPTION
**Pull request by xiaoxubeii**
Signed-off-by: xiaoxubeii xiaoxubeii@gmail.com

### Ⅰ. Describe what this PR did
Fix that cannot update host bridge when restart pouchd daemon

### Ⅱ. Does this pull request fix one issue?

fixes #2128.

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
YES.

### Ⅳ. Describe how to verify it

```
$ systemctl start pouch
$ ip a show p0
119: p0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN
    link/ether 02:42:c0:a8:05:01 brd ff:ff:ff:ff:ff:ff
    inet 192.168.5.1/24 brd 192.168.5.255 scope global p0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:c0ff:fea8:501/64 scope link
       valid_lft forever preferred_lft forever

$ cat  <<EOF > /etc/pouch/config.json
{
  "network-config": {
    "bridge-config": {
      "bridge-name": "p0",
      "bip": "172.20.0.1/24",
      "mtu": 1500,
      "default-gateway": "172.20.0.1"
    }
  }
}
EOF

$ system restart pouch && ip a show p0
120: p0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN
    link/ether 02:42:ac:14:00:01 brd ff:ff:ff:ff:ff:ff
    inet 172.20.0.1/24 brd 172.20.0.255 scope global p0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe14:1/64 scope link
       valid_lft forever preferred_lft forever
```
### Ⅴ. Special notes for reviews
NONE.

### CODE REVIEW

File: network/mode/bridge/bridge.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/alibaba/pouch/network"
 	"github.com/alibaba/pouch/pkg/errtypes"
 
+	"github.com/alibaba/pouch/pkg/utils"
```
Comment (rudyfly) [inline network/mode/bridge/bridge.go]: move this line to previous row

**Comment by rudyfly** [comment]
What scene to modify bridge ip address?
On our server, bridge is already exist and can't to remove it, by your method will remove bridge that have been exist. It isn't applicable to our server.

**Comment by xiaoxubeii** [comment]
@rudyfly I am confused by the bridge process, because in https://github.com/alibaba/pouch/blob/master/network/mode/bridge/bridge.go#L27: 
```
// New is used to initialize bridge network.
func New(ctx context.Context, config network.BridgeConfig, manager mgr.NetworkMgr) error {
	// TODO: support ipv6.

	// clear exist bridge network
	if err := manager.Remove(ctx, "bridge"); err != nil {
		if !errtypes.IsNotfound(err) {
			return err
		}
	}
        ...
        // init host bridge network.
	br, err := initBridgeDevice(bridgeName)
	if err != nil {
		return err
	}

...
```
It indicates that the bridge can be changed or created by configuration, but in `initBridgeDevice`, it uses the default. That's why i post this pr.

**Comment by allencloud** [comment]
Any update for this pull request. @rudyfly

**Comment by xiaoxubeii** [comment]
@rudyfly @allencloud ping for next move.

File: network/mode/bridge/bridge.go [hunk 2]
```diff
@@ -148,15 +115,18 @@ func New(ctx context.Context, config network.BridgeConfig, manager mgr.NetworkMg
 	return err
 }
 
-func initBridgeDevice(name string) (netlink.Link, error) {
+func initBridgeDevice(name string, ipNet *net.IPNet) (netlink.Link, error) {
 	br, err := netlink.LinkByName(name)
-	if err == nil && br != nil {
-		return br, nil
+	if err == nil && br != nil { // we need delete old link
+		err := netlink.LinkDel(br)
```
Comment (rudyfly) [inline network/mode/bridge/bridge.go]: Add a condition, if the exist bridge's ip is same with config bip, it no need to delete old bridge. And check the veth-pair is attached on the exist bridge, this case can't delete bridge device.

**Comment by xiaoxubeii** [comment]
@rudyfly Done

**Comment by xiaoxubeii** [comment]
@rudyfly Done

**Comment by rudyfly** [comment]
LGTM, Thanks for your contribution. @xiaoxubeii

**Comment by allencloud** [comment]
Thanks for your contribution. @xiaoxubeii

**Pull request merged**